Objects/bytesobject.c

```diff
@@ -2810,82 +2811,46 @@ _PyBytes_FromBuffer(PyObject *x)
 }
 
 static PyObject*
-_PyBytes_FromList(PyObject *x)
+_PyBytes_FromSequence(PyObject *x)
 {
-    Py_ssize_t i, size = PyList_GET_SIZE(x);
-    Py_ssize_t value;
-    char *str;
-    PyObject *item;
-    _PyBytesWriter writer;
-
-    _PyBytesWriter_Init(&writer);
-    str = _PyBytesWriter_Alloc(&writer, size);
-    if (str == NULL)
+    Py_ssize_t size = PySequence_Fast_GET_SIZE(x);
+    PyObject *bytes = _PyBytes_FromSize(size, 0);
+    if (bytes == NULL) {
         return NULL;
-    writer.overallocate = 1;
-    size = writer.allocated;
-
-    for (i = 0; i < PyList_GET_SIZE(x); i++) {
-        item = PyList_GET_ITEM(x, i);
-        Py_INCREF(item);
-        value = PyNumber_AsSsize_t(item, NULL);
-        Py_DECREF(item);
-        if (value == -1 && PyErr_Occurred())
-            goto error;
-
-        if (value < 0 || value >= 256) {
-            PyErr_SetString(PyExc_ValueError,
-                            "bytes must be in range(0, 256)");
-            goto error;
-        }
-
-        if (i >= size) {
-            str = _PyBytesWriter_Resize(&writer, str, size+1);
-            if (str == NULL)
-                return NULL;
-            size = writer.allocated;
-        }
-        *str++ = (char) value;
     }
-    return _PyBytesWriter_Finish(&writer, str);
-
-  error:
-    _PyBytesWriter_Dealloc(&writer);
-    return NULL;
-}
-
-static PyObject*
-_PyBytes_FromTuple(PyObject *x)
-{
-    PyObject *bytes;
-    Py_ssize_t i, size = PyTuple_GET_SIZE(x);
-    Py_ssize_t value;
-    char *str;
-    PyObject *item;
-
-    bytes = PyBytes_FromStringAndSize(NULL, size);
-    if (bytes == NULL)
-        return NULL;
-    str = ((PyBytesObject *)bytes)->ob_sval;
-
-    for (i = 0; i < size; i++) {
-        item = PyTuple_GET_ITEM(x, i);
-        value = PyNumber_AsSsize_t(item, NULL);
-        if (value == -1 && PyErr_Occurred())
+    char *str = PyBytes_AS_STRING(bytes);
+    PyObject *const *items = PySequence_Fast_ITEMS(x);
+    Py_BEGIN_CRITICAL_SECTION_SEQUENCE_FAST(x);
+    for (Py_ssize_t i = 0; i < size; i++) {
+        if (!PyLong_Check(items[i])) {
```

Comment:
`PyLong_CheckExact` probably fits better here, and is faster!

Reply:
(The previous code wasn't doing an exact check so we shouldn't change it)

Reply:
Well, it wouldn't be a breaking change, just a performance loss for the (very niche!) set of cases that use special ints. I'm also slightly worried that non-exact ints might have some nasty side effects that we aren't anticipating here (e.g. can they mess with the `Py_ssize_t` value?)

Reply:
They can mess with `Py_ssize_t` values, but the problem is wider. For example, `np.int32()` values can be passed to `bytes([...])` but without this check, we will impact performances of numpy-related code which is something we don't want to.

Reply:
Is casting to bytes a common thing to do with numpy integers, or is that speculation? (I see your point, I'm just sort of gauging what the cost-benefit would be here.)

Reply:
I'd say yes, if we're considering serialization or introspection. I imagine that we can have such things when people are working with images because their arrays won't necessarily be pure Python lists but `np.ndarray` objects which may have non-primitive data types. I'm not aware of a wild opensource usage though but IMO, since we want to improve performance overall,  we shouldn't penalize existing users. If we can slightly decrease our performance gain so to have a stable result, it's good.

Nonetheless, if we want to fallback to a slow path for non-exact ints, benchmarks should show how performances are impacted (a simple `timeit` could be sufficient but I'd be more comfortable with a much precise benchmark).